src/ray/common/function_descriptor.h

```diff
@@ -0,0 +1,181 @@
+#ifndef RAY_CORE_WORKER_FUNCTION_DESCRIPTOR_H
+#define RAY_CORE_WORKER_FUNCTION_DESCRIPTOR_H
+
+#include <string>
+
+#include "ray/common/grpc_util.h"
+#include "ray/protobuf/common.pb.h"
+
+namespace ray {
+/// See `common.proto` for definition of `FunctionDescriptor` oneof type.
+using FunctionDescriptorType = rpc::FunctionDescriptor::FunctionDescriptorCase;
+/// Wrap a protobuf message.
+class FunctionDescriptorInterface : public MessageWrapper<rpc::FunctionDescriptor> {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  FunctionDescriptorInterface(rpc::FunctionDescriptor message)
+      : MessageWrapper(std::move(message)) {}
+
+  ray::FunctionDescriptorType Type() const {
+    return message_->function_descriptor_case();
+  }
+
+  virtual size_t Hash() const = 0;
+
+  virtual std::string ToString() const = 0;
+
+  template <typename Subtype>
+  Subtype *As() {
+    return reinterpret_cast<Subtype *>(this);
+  }
+};
+
+class DriverFunctionDescriptor : public FunctionDescriptorInterface {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  explicit DriverFunctionDescriptor(rpc::FunctionDescriptor message)
+      : FunctionDescriptorInterface(std::move(message)) {
+    RAY_CHECK(message_->function_descriptor_case() ==
+              ray::FunctionDescriptorType::kDriverFunctionDescriptor);
+  }
+
+  virtual size_t Hash() const {
+    return std::hash<int>()(ray::FunctionDescriptorType::kDriverFunctionDescriptor);
+  }
+
+  virtual std::string ToString() const { return "{type=DriverFunctionDescriptor}"; }
+};
+
+class JavaFunctionDescriptor : public FunctionDescriptorInterface {
+ public:
+  /// Construct from a protobuf message object.
+  /// The input message will be **copied** into this object.
+  ///
+  /// \param message The protobuf message.
+  explicit JavaFunctionDescriptor(rpc::FunctionDescriptor message)
+      : FunctionDescriptorInterface(std::move(message)) {
+    RAY_CHECK(message_->function_descriptor_case() ==
+              ray::FunctionDescriptorType::kJavaFunctionDescriptor);
+    typed_message_ = &(message_->java_function_descriptor());
+  }
+
+  virtual size_t Hash() const {
+    return std::hash<int>()(ray::FunctionDescriptorType::kJavaFunctionDescriptor) ^
```

Comment:
Is it only used once?
In fact, we can store hash value in member property if it's used in many code pathes.

Reply:
I have tested that a dedicated hash value field will does make code significantly faster.